src/java.base/share/classes/java/lang/runtime/TemplateRuntime.java

```diff
@@ -0,0 +1,224 @@
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.StringTemplate.Processor;
+import java.lang.StringTemplate.Processor.Linkage;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+import jdk.internal.access.JavaTemplateAccess;
+import jdk.internal.access.SharedSecrets;
+import jdk.internal.javac.PreviewFeature;
+
+/**
+ * Manages string template bootstrapping and creation. These methods may be used, for example,
+ * by Java compiler implementations to implement the bodies of methods for {@link StringTemplate}
+ * objects.
+ * <p>
+ * The {@link TemplateRuntime#newLargeStringTemplate} bootstrap method is used to create
+ * {@link StringTemplate StringTemplates} that have more than
+ * {@link java.lang.invoke.StringConcatFactory#MAX_INDY_CONCAT_ARG_SLOTS} values.
+ * <p>
+ * The {@link TemplateRuntime#newStringTemplate} bootstrap method is used to create
+ * optimized {@link StringTemplate StringTemplates}.
+ * <p>
+ * The {@link TemplateRuntime#newLargeStringTemplate} bootstrap method is used
+ * to bind to specialized processors that implement {@link Linkage}.
+ *
+ * @since 21
+ */
+@PreviewFeature(feature=PreviewFeature.Feature.STRING_TEMPLATES)
+public final class TemplateRuntime {
+    private static final JavaTemplateAccess JTA = SharedSecrets.getJavaTemplateAccess();
+
+    /**
+     * {@link MethodHandle} to {@link TemplateRuntime#defaultProcess}.
+     */
+    private static final MethodHandle DEFAULT_PROCESS_MH;
+
+    /**
+     * {@link MethodHandle} to {@link TemplateRuntime#newTrustedStringTemplate}.
+     */
+    private static final MethodHandle NEW_TRUSTED_STRING_TEMPLATE;
+
+    /**
+     * Initialize {@link MethodHandle MethodHandles}.
+     */
+    static {
+        try {
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+
+            MethodType mt = MethodType.methodType(Object.class,
+                    List.class, Processor.class, Object[].class);
+            DEFAULT_PROCESS_MH =
+                lookup.findStatic(TemplateRuntime.class, "defaultProcess", mt);
+
+            mt = MethodType.methodType(StringTemplate.class, String[].class, Object[].class);
+            NEW_TRUSTED_STRING_TEMPLATE =
+                lookup.findStatic(StringTemplateImplFactory.class, "newTrustedStringTemplate", mt);
+        } catch (ReflectiveOperationException ex) {
+            throw new AssertionError("string bootstrap fail", ex);
+        }
+    }
+
+    /**
+     * Private constructor.
+     */
+    private TemplateRuntime() {
+        throw new AssertionError("private constructor");
+    }
+
+    /**
+     * String template bootstrap method for creating large string templates.
+     * The non-static arguments are the fragments list and values list.
+     *
+     * @param lookup          method lookup
+     * @param name            method name
+     * @param type            method type
```

Comment:
Maybe mention that the name is ignored, and the type must be convertible from `(String[], Object[])StringTemplate`? The specification says "fragments list" and "values list", which are more accurately described as arrays.

Also, does "large" string template mean there are more than 250~ fragments that the bootstrap method arguments can't fit, or does it mean dynamic number of fragments? Might be worth mentioning that as well.

Reply:
Updating comments